LLama/Model/ModelManager.cs

```diff
@@ -0,0 +1,309 @@
+    {
+        GetModelsFromDirectories(directory);
+    }
+
+    /// <inheritdoc />
+    public bool RemoveDirectory(string directory)
+    {
+        return _availableModels.Remove(Path.GetFullPath(directory));
+    }
+
+    /// <inheritdoc />
+    public void RemoveAllDirectories()
+    {
+        _availableModels.Clear();
+    }
+
+    /// <inheritdoc />
+    public IEnumerable<ModelFileMetadata> GetAvailableModelsFromDirectory(string directory)
+    {
+        var dirPath = Path.GetFullPath(directory);
+        return _availableModels.TryGetValue(dirPath, out var dirModels)
+            ? dirModels
+            : [];
+    }
+
+    /// <inheritdoc />
+    public bool TryGetModelFileMetadata(string fileName, out ModelFileMetadata modelMeta)
+    {
+        var filePath = Path.GetFullPath(fileName);
+        modelMeta = ModelFileList.FirstOrDefault(f => f.FilePath == filePath)!;
+        return modelMeta != null;
+    }
+
+    /// <inheritdoc />
+    public IEnumerable<LLamaWeights> GetLoadedModels()
+    {
+        return _loadedModelCache.Values;
+    }
+
+    /// <inheritdoc />
+    public bool TryGetLoadedModel(string modelId, out LLamaWeights model)
+    {
+        return _loadedModelCache.TryGetValue(modelId, out model!);
+    }
+
+    /// <inheritdoc />
+    public async Task<LLamaWeights> LoadModel(string modelPath,
+        Action<ModelParams>? modelConfigurator = null!,
+        string modelId = "",
+        CancellationToken cancellationToken = default)
+    {
+        // Configure model params
+        var modelParams = new ModelParams(modelPath);
+        modelConfigurator ??= DefaultModelConfigurator;
+        modelConfigurator.Invoke(modelParams);
+
+        // load and cache
+        var model = await LLamaWeights.LoadFromFileAsync(modelParams, cancellationToken);
+        if (string.IsNullOrWhiteSpace(modelId))
+        {
+            modelId = model.ModelName;
+        }
+        _loadedModelCache.Add(modelId, model);
+        return model;
+    }
+
+    /// <summary>
+    /// Updates the passed in model params with default values
+    /// </summary>
+    /// <param name="modelParams"></param>
+    public static void DefaultModelConfigurator(ModelParams modelParams)
+    {
+        // Reasonable defaults
+        modelParams.GpuLayerCount = 12;
+        modelParams.Seed = 1337u;
+        modelParams.ContextSize = 2048;
+        modelParams.Encoding = Encoding.UTF8;
+
+        modelParams.UseMemoryLock = true;
+        modelParams.UseMemorymap = true;
+
+        // auto detect
+        modelParams.Threads = null!;
+        modelParams.BatchThreads = null!;
+    }
+
+    /// <inheritdoc />
+    public bool UnloadModel(string modelId)
+    {
+        if (TryGetLoadedModel(modelId, out var model))
+        {
+            model.Dispose();
+            return _loadedModelCache.Remove(modelId);
+        }
+        return false;
+    }
+
+    /// <inheritdoc />
+    public void UnloadAllModels()
```

Comment:
This entire class should probably be `IDisposable` since it owns all of the `LLamaWeights` objects (with the Dispose implementation calling `UnloadAllModels()`

Reply:
I internally debated on this a bit and ultimately decided against it and figured the caller should be responsible for the model and dispose of it themselves or call unload explicitly. Do you feel strongly about this class being disposable itself? 

Main use case would be something like 
`using var model ModelCache.LoadModel(...);`

where they wouldn't call Unload explicitly. To your point above though, `TryGetLoadedModel` should guard against externally disposed models. and same with the call to load, if it's already in the cache.

Reply:
> I internally debated on this a bit and ultimately decided against it and figured the caller should be responsible for the model

I think the important thing is to be very clear about resource management, that's probably the most important problem that LLamaSharp has to solve in wrapping an unmanaged set of resources up for use in a managed language!

The problem with the cache not having ownership of the weights becomes _very_ unclear. If I call `LoadModel` from some code then it might be reasonable to assume I own those weights and should dispose them when done... but then the entire concept of a shared cache is useless (I unloaded the weights, so no one else can use them) or even harmful (I dispose the shared weights while some other bit of code is using them).

---

While typing up the above I did have another idea:

The lowest level thing that holds ownership of any resource is always a `SafeHandle` (e.g. [`SafeLlamaModelHandle`](https://github.com/SciSharp/LLamaSharp/blob/master/LLama/Native/SafeLlamaModelHandle.cs) for weights. A `SafeHandle` is a **reference counted** owner of a resource, caling dispose on a safe handle won't actually dispose the underlying resource until the count goes to zero.

We use this in `SafeLLamaContextHandle`. The context adds 1 to the weights ref count until the context is disposed. That way this code actually works:

```csharp
var weights = LoadWeightsFromSomewhere();
var ctx = weights.CreateContext();
weights.Dispose();
ctx.DoSomething(); // <-- this is fine, even thought it references the weights we just "disposed"
ctx.Dispose(); // <-- _now_ the weights are actually disposed
```

We could do the same thing in `LLamaWeights`. The cache can internally contain `SafeLLamaModelHandle` objects (which it owns, and disposes when the cache is disposed). When a request is made for some weights they get wrapped up in a new `LLamaWeights` objects (which increments the count), the caller unambiguously owns this object.

Note: if this is all sounds a bit scary and you'd rather not implement it I understand, I'm happy to have a look at it in a separate PR which we can merge first and then rebase this one onto.

Reply:
Doesn't sound scary, let me explore a bit 😄

Reply:
LLamaContext should serve as a good example of how to do it

Reply:
Alright, let me know what you think of the latest changes